core/blockchain.go

```diff
@@ -2018,6 +2029,105 @@ func (bc *BlockChain) update() {
 	}
 }
 
+// updateTxIndices is responsible for the construction and deletion of the
+// transaction index.
+//
+// User can use flag `txlookuplimit` to specify a "recentness" block, below
+// which ancient tx indices get deleted. If `txlookuplimit` is 0, it means
+// all tx indices will be reserved.
+//
+// The user can adjust the txlookuplimit value for each launch, Geth will
+// automatically construct the missing indices and delete the extra indices.
+func (bc *BlockChain) updateTxIndices() {
+	var (
+		done   chan struct{} // Non-nil if background unindexing or reindexing routine is active.
+		headCh = make(chan ChainHeadEvent)
+	)
+	sub := bc.SubscribeChainHeadEvent(headCh)
+	defer func() {
+		if sub != nil {
+			sub.Unsubscribe()
+		}
+	}()
+	// initialiseIndices inits txlookup indices into the database.
+	// If there already exists some indices, this function will find
+	// the oldest block which has been indexed and start indexing from
+	// this point.
+	initialiseIndices := func(head uint64) {
```

Comment:
Do we need this `initialiseIndices`? If we fix the receipt writer in fast sync to actually push the tx indexes too for recent blocks, then I think that would already make this part unneeded. Maybe a fully empty chain would need some sanity checks, but not heavy reindexing.